### ISSUE TITLE
certificate chain filename reporting regression

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8222]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#8222|
|Type|release blocker: regression|
|Created|2016-02-28 14:43:37Z|
|Branch|https://github.com/twisted/twisted/tree/cert-chain-reg-8222|

This program:

```
from twisted.internet import reactor
from twisted.internet.protocol import Factory
from twisted.internet.endpoints import serverFromString

with open("empty.pem", "wt") as e:
    pass

server = serverFromString(reactor, "ssl:12345:extraCertChain=empty.pem")
server.listen(Factory())
reactor.run()
```

produces this output on master@HEAD

```
Traceback (most recent call last):
  File "twisted-regression.py", line 8, in <module>
    server = serverFromString(reactor, "ssl:12345:extraCertChain=empty.pem")
  File "/home/jean-paul/Projects/twisted/master/twisted/internet/endpoints.py", line 1523, in serverFromString
    return _serverFromStringLegacy(reactor, description, _NO_DEFAULT)
  File "/home/jean-paul/Projects/twisted/master/twisted/internet/endpoints.py", line 1436, in _serverFromStringLegacy
    nameOrPlugin, args, kw = _parseServer(description, None, default)
  File "/home/jean-paul/Projects/twisted/master/twisted/internet/endpoints.py", line 1427, in _parseServer
    return (endpointType.upper(),) + parser(factory, *args[1:], **kw)
  File "/home/jean-paul/Projects/twisted/master/twisted/internet/endpoints.py", line 1152, in _parseSSL
    "certificates in PEM format." % (extraCertChain,)
ValueError: Specified chain file '' doesn't contain any valid certificates in PEM format.
```

It used to produce this output:

```
Traceback (most recent call last):
  File "twisted-regression.py", line 8, in <module>
    server = serverFromString(reactor, "ssl:12345:extraCertChain=empty.pem")
  File "/usr/lib/python2.7/dist-packages/twisted/internet/endpoints.py", line 1465, in serverFromString
    return _serverFromStringLegacy(reactor, description, _NO_DEFAULT)
  File "/usr/lib/python2.7/dist-packages/twisted/internet/endpoints.py", line 1393, in _serverFromStringLegacy
    nameOrPlugin, args, kw = _parseServer(description, None, default)
  File "/usr/lib/python2.7/dist-packages/twisted/internet/endpoints.py", line 1384, in _parseServer
    return (endpointType.upper(),) + parser(factory, *args[1:], **kw)
  File "/usr/lib/python2.7/dist-packages/twisted/internet/endpoints.py", line 1121, in _parseSSL
    "certificates in PEM format." % (extraCertChain,)
ValueError: Specified chain file 'empty.pem' doesn't contain any valid certificates in PEM format.
```

<details><summary>Searchable metadata</summary>

```
trac-id__8222 8222
type__release_blocker__regression release blocker: regression
reporter__exarkun exarkun
priority__normal normal
milestone__None None
branch__branches_cert_chain_reg_8222 branches/cert-chain-reg-8222
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1456670617172227 1456670617172227
changetime__1456988147541238 1456988147541238
version__None None
owner__hawkowl hawkowl

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

It seems probable that this was introduced by the resolution to [#7982](https://github.com/twisted/twisted/issues/7982).

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_2"></a>|@hawkowl set owner to @hawkowl|
|-|-|

Yep, it was -- the meaning of ```extraCertChain``` was changed and no tests picked up the differing error message. 

I broke it, so I'll fix it ;)

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_3"></a>|@hawkowl commented|
|-|-|

(In [46880]) Branching to cert-chain-reg-8222.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_4"></a>|@hawkowl removed owner|
|-|-|

Okay, I think this should fix it. Builders are looking green so far, please review.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_5"></a>|@adiroiban set owner to @hawkowl|
|-|-|

Thanks for the patch.

Changes looks good.

That variable name abuse was ugly. My bad for not detecting it at the initial review.

Since this regression was introduced in 15.4, it is OK for this branch to only have a .misc release notes? I was expecting a .bugfix

Please check my comments and merge.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_6"></a>|@hawkowl set status to `closed`|
|-|-|

(In [46886]) Merge cert-chain-reg-8222: Fix serverFromString's SSL parsing having an incorrect error message when an empty chain file is given

Author: hawkowl
Reviewer: adiroiban
Fixes: [#8222](https://github.com/twisted/twisted/issues/8222)
